url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/Node.java

```diff
@@ -0,0 +1,1030 @@
+
+		m_totalDataLength -= m_pDataLength[index];
+		m_pData[index] = null;
+
+		if (m_children > 1 && index != m_children - 1) {
+			//replace the target entry with last enrty, then m_children--
+			m_pDataLength[index] = m_pDataLength[m_children - 1];
+			m_pData[index] = m_pData[m_children - 1];
+			m_pData[m_children - 1] = null;
+			m_pMBR[index] = m_pMBR[m_children - 1];
+			m_pMBR[m_children - 1] = null;
+			m_pIdentifier[index] = m_pIdentifier[m_children - 1];
+		}
+
+		m_children--;
+
+		if (m_children == 0) {
+			// means that no child remains
+			m_nodeMBR = (Region) m_pTree.m_infiniteRegion.clone();
+		} else if (touches) {
+			for (int cDim = 0; cDim < m_pTree.m_dimension; cDim++) {
+				m_nodeMBR.m_pLow[cDim] = Double.POSITIVE_INFINITY;
+				m_nodeMBR.m_pHigh[cDim] = Double.NEGATIVE_INFINITY;
+
+				for (int cChild = 0; cChild < m_children; cChild++) {
+					m_nodeMBR.m_pLow[cDim] = Math.min(m_nodeMBR.m_pLow[cDim], m_pMBR[cChild].m_pLow[cDim]);
+					m_nodeMBR.m_pHigh[cDim] = Math.max(m_nodeMBR.m_pHigh[cDim], m_pMBR[cChild].m_pHigh[cDim]);
+				}
+			}
+		}
+	}
+
+	/**
+	 * insert entry and save it
+	 * 
+	 * @param pData
+	 * @param mbr
+	 * @param id
+	 * @param pathBuffer
+	 * @param overflowTable
+	 * @return
+	 */
+	protected boolean insertData(byte[] pData, Region mbr, int id, Stack<Integer> pathBuffer, boolean[] overflowTable, ArrayList<Integer> changedNodeID) {
+		if (m_children < m_capacity) {
+			boolean adjusted = false;
+			boolean b = m_nodeMBR.contains(mbr);
+
+			insertEntry(pData, mbr, id);
+			m_pTree.writeNode(this);//save this change
+
+			if (!b && !pathBuffer.empty()) {
+				int cParent = (pathBuffer.pop()).intValue();
+				Index p = (Index) m_pTree.readNode(cParent);
+				p.adjustTree(this, pathBuffer);
+				adjusted = true;
+			}
+			if (changedNodeID != null)
+				changedNodeID.add(m_identifier);
+			return adjusted;
+		} else if (m_pTree.m_treeVariant == SpatialIndex.RtreeVariantRstar && !pathBuffer.empty() && overflowTable[m_level] == false) {
+			overflowTable[m_level] = true;
+
+			ArrayList<Integer> vReinsert = new ArrayList<Integer>(), vKeep = new ArrayList<Integer>();
+			reinsertData(pData, mbr, id, vReinsert, vKeep);
+
+			int lReinsert = vReinsert.size();
+			int lKeep = vKeep.size();
+
+			byte[][] reinsertdata = new byte[lReinsert][];
+			Region[] reinsertmbr = new Region[lReinsert];
+			int[] reinsertid = new int[lReinsert];
+			int[] reinsertlen = new int[lReinsert];
+			byte[][] keepdata = new byte[m_capacity + 1][];
+			Region[] keepmbr = new Region[m_capacity + 1];
+			int[] keepid = new int[m_capacity + 1];
+			int[] keeplen = new int[m_capacity + 1];
+
+			int cIndex;
+
+			for (cIndex = 0; cIndex < lReinsert; cIndex++) {
+				int i = (vReinsert.get(cIndex)).intValue();
+				reinsertlen[cIndex] = m_pDataLength[i];
+				reinsertdata[cIndex] = m_pData[i];
+				reinsertmbr[cIndex] = m_pMBR[i];
+				reinsertid[cIndex] = m_pIdentifier[i];
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 297:303, 305:311.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.